lms/security.py

```diff
@@ -120,52 +144,58 @@ def identity(self, request):
 
         return Identity("", [])
 
-    def permits(self, request, context, permission):
-        return _permits(self, request, context, permission)
+    def permits(self, request, _context, permission):
+        return _permits(self.identity(request), permission)
 
 
-def _permits(policy, request, _context, permission):
-    identity = policy.identity(request)
+def _permits(identity, permission):
     if identity and permission in identity.permissions:
         return Allowed("allowed")
 
     return Denied("denied")
 
 
-def _get_lti_user(request) -> LTIUser:
+@lru_cache(maxsize=1)
+def get_lti_user_from_launch_params(request) -> LTIUser:
+    if "id_token" in request.params:
+        return LTI13AuthSchema(request).lti_user()
+
+    return LTI11AuthSchema(request).lti_user()
```

Comment:
Do these schemas ever return `None`? Should it be `Optional[LTIUser]`?

Reply:
No, they either raise or return the LTIUser.